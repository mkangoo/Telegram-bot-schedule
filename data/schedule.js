export const dataBase = [
	{
		day: 'Monday',
		events: [
			{
				start: '2024-01-22T08:15:00',
				end: '2024-01-22T09:45:00',
				title: 'Организационное сопровождение стартапа',
				lection: [2, 3, 4, 5, 6, 7],
				practical: [8, 9],
				description: 'Дегтярёва В.В.',
				audience: 'ЛК - 538',
			},
			{
				start: '2024-01-22T09:55:00',
				end: '2024-01-22T11:25:00',
				title: 'Организационное сопровождение стартапа',
				lection: [],
				practical: [2, 3, 4, 5, 6, 7, 8, 9],
				description: 'Дегтярёва В.В.',
				audience: 'ЛК - 538',
			},
			{
				start: '2024-01-22T11:35:00',
				end: '2024-01-22T13:05:00',
				title: 'Разработка бизнес-плана инновационного проекта',
				lection: [2, 3, 4, 5, 6, 7],
				practical: [8, 9],
				description: 'Дзюрдзя О.А.',
				audience: 'ЛК - 407',
			},
			{
				start: '2024-01-22T13:45:00',
				end: '2024-01-22T15:15:00',
				title: 'Разработка бизнес-плана инновационного проекта',
				lection: [],
				practical: [2, 4, 6, 8],
				description: 'Дзюрдзя О.А.',
				audience: 'ЛК - 408',
			},
		],
	},
	{
		day: 'Tuesday',
		events: [
			{
				start: '2024-01-22T08:15:00',
				end: '2024-01-22T15:15:00',
				title: 'Проектное обучение',
				lection: [],
				practical: [2, 3, 4, 5, 6, 7, 8, 9],
			},
		],
	},
	{
		day: 'Wednesday',
		events: [
			{
				start: '2024-01-22T09:55:00',
				end: '2024-01-22T11:25:00',
				title: 'Управление развитием инновационного продукта',
				lection: [2, 3, 4, 5, 6, 7],
				practical: [8, 9],
				description: 'Дегтярёва В.В.',
				audience: 'ЛК - 402',
			},
			{
				start: '2024-01-22T11:35:00',
				end: '2024-01-22T13:05:00',
				title: 'Управление развитием инновационного продукта',
				lection: [],
				practical: [2, 3, 4, 5, 6, 7, 8, 9],
				description: 'Дегтярёва В.В.',
				audience: 'ЛК - 402',
			},
			{
				start: '2024-01-22T13:45:00',
				end: '2024-01-22T15:15:00',
				title: 'Управление развитием инновационного продукта',
				lection: [],
				practical: [3, 5, 7],
				description: 'Дегтярёва В.В.',
				audience: 'ЛК - 402',
			},
		],
	},
	{
		day: 'Thursday',
		events: [
			{
				start: '2024-01-22T08:15:00',
				end: '2024-01-22T09:45:00',
				title: 'Проектирование инвестиционного меморандума',
				lection: [2, 3, 4, 5, 6, 7],
				practical: [8, 9],
				description: 'Бигачева Е.Н.',
				audience: 'ЛК - 407',
			},
			{
				start: '2024-01-22T09:55:00',
				end: '2024-01-22T11:25:00',
				title: 'Проектирование инвестиционного меморандума',
				lection: [],
				practical: [2, 3, 4, 5, 6, 7, 8, 9],
				description: 'Бигачева Е.Н.',
				audience: 'ЛК - 407',
			},
			{
				start: '2024-01-22T11:35:00',
				end: '2024-01-22T13:05:00',
				title: 'Разработка бизнес-плана инновационного проекта',
				lection: [],
				practical: [3, 5, 7, 9],
				description: 'Дзюрдзя О.А.',
				audience: 'ЛК-402',
			},
			{
				start: '2024-01-22T11:35:00',
				end: '2024-01-22T13:05:00',
				title: 'Организационное сопровождение стартапа',
				lection: [],
				practical: [4, 6],
				description: 'Дегтярева В.В',
				audience: 'ЛК-411',
			},
			{
				start: '2024-01-22T11:35:00',
				end: '2024-01-22T13:05:00',
				title: 'Проектирование инвестиционного меморандума',
				lection: [],
				practical: [8, 10],
				description: 'Бигачева Е.Н.',
				audience: 'ЛК-407',
			},
			{
				start: '2024-01-22T13:45:00',
				end: '2024-01-22T15:15:00',
				title: 'Разработка бизнес-плана инновационного проекта',
				lection: [],
				practical: [3, 5],
				description: 'Дзюрдзя О.А.',
				audience: 'ЛК-402',
			},
			{
				start: '2024-01-22T13:45:00',
				end: '2024-01-22T15:15:00',
				title: 'Проектирование инвестиционного меморандума',
				lection: [],
				practical: [8],
				description: 'Бигачева Е.Н.',
				audience: 'ЛК-407',
			},
		],
	},
	{
		day: 'Friday',
		events: [
			{
				start: '2024-01-22T08:15:00',
				end: '2024-01-22T09:45:00',
				title: 'Управление рисками инновационной деятельности',
				lection: [1, 2, 3, 4, 5, 6],
				practical: [7, 8],
				description: 'Сергеева К.Н.',
				audience: 'ЛК - 507',
			},
			{
				start: '2024-01-22T09:55:00',
				end: '2024-01-22T11:25:00',
				title: 'Управление рисками инновационной деятельности',
				lection: [],
				practical: [1, 2, 3, 4, 5, 6, 7, 8],
				description: 'Сергеева К.Н.',
				audience: 'ЛК - 507',
			},
			{
				start: '2024-01-22T11:35:00',
				end: '2024-01-22T13:05:00',
				title: 'Управление рисками инновационной деятельности',
				lection: [],
				practical: [2, 4],
				description: 'Сергеева К.Н.',
				audience: 'ЛК - 411',
			},
		],
	},
]
