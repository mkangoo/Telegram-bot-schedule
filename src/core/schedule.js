// eslint-disable-next-line import/no-cycle
import { orderedWeekDays, NO_LESSONS_MESSAGE } from '../utils/constant.js'
import { getLessonType } from '../utils/lessonType.js'
import { getWeekNumber } from '../utils/weekNumber.js'
import { dataBase } from '../../data/schedule.js'

export const getScheduleOutput = (lessonType, { start, end, title, description, audience }) => {
	// Extract start and end times
	const startTime = start.slice(11, 16)
	const endTime = end.slice(11, 16)

	const lessonTitle = title || ''
	const lessonDescription = description || ''
	const audienceInfo = audience || ''

	// Format lesson information
	return `\n\n‚è∞ ${startTime} - ${endTime}\n${lessonType}:\n<b>${lessonTitle}</b>\n${lessonDescription}\n${audienceInfo}`
}

export function getLessonsForDay(targetDay, scheduleData, shouldShiftWeek, weekNumber = getWeekNumber(new Date())) {
	const lessonToday = scheduleData.find(({ day, events }) => day === targetDay && events)
	if (!lessonToday) return NO_LESSONS_MESSAGE

	if (shouldShiftWeek) weekNumber += 1

	const formattedLessons = lessonToday.events
		.map(event => {
			const lessonType = getLessonType(event, weekNumber)
			return lessonType ? getScheduleOutput(lessonType, event) : null
		})
		.filter(info => info !== null)
		.join(' ')

	return formattedLessons || NO_LESSONS_MESSAGE
}

export const getFullSchedule = shiftWeek => {
	const scheduleWeek = orderedWeekDays
		.filter(day => day !== 'Sunday' && day !== 'Saturday')
		.map(day => {
			const scheduleDay = getLessonsForDay(day, dataBase, shiftWeek).trim()
			return `\nüõë <b>${day}</b>:\n\n ${scheduleDay}\n`
		})
	return scheduleWeek.join('')
}

const getDayIndex = () => {
	return {
		today() {
			const date = new Date()
			const currentHours = date.getUTCHours()
			// Time zone offset in Moscow
			date.setUTCHours(currentHours + 3)
			const currentDay = date.getDay()
			const dayIndex = currentDay % 7
			return dayIndex % 7
		},
		tomorrow() {
			return this.today() + 1
		},
	}
}

const holidayStatus = () => {
	const dayIndex = getDayIndex().today()
	return orderedWeekDays[dayIndex] === 'Sunday'
}

export const getTodaySchedule = () => {
	const dayIndex = getDayIndex().today()
	return getLessonsForDay(orderedWeekDays[dayIndex], dataBase)
}

export const getTomorrowSchedule = () => {
	const dayIndex = getDayIndex().tomorrow()
	return getLessonsForDay(orderedWeekDays[dayIndex], dataBase, { raw: holidayStatus() })
}
