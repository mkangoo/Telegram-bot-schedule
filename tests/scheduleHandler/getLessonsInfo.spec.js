import { dataBase } from '../../data/schedule.js'
import { getLessonsForDay } from '../../src/core/schedule.js'

describe('Works when today is weekend day', () => {
	it('Returns no lessons message for Sunday', () => {
		expect(getLessonsForDay('Sunday', dataBase, { weekOffset: false })).toEqual('<b>üéâ –ó–∞–Ω—è—Ç–∏–π –Ω–µ—Ç, –º–æ–∂–Ω–æ –æ—Ç–¥—ã—Ö–∞—Ç—å.</b>')
	})
	it('Returns no lessons for Saturday', () => {
		expect(getLessonsForDay('Saturday', dataBase, { weekOffset: false })).toEqual('<b>üéâ –ó–∞–Ω—è—Ç–∏–π –Ω–µ—Ç, –º–æ–∂–Ω–æ –æ—Ç–¥—ã—Ö–∞—Ç—å.</b>')
	})
})
describe('Works when today is weekday', () => {
	const WEEK_DAY = 'Wednesday'
	const MOCK_SCHEDULE_DATA = [
		{
			day: 'Wednesday',
			events: [
				{
					start: '2024-01-22T09:55:00',
					end: '2024-01-22T11:25:00',
					title: '–£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ —Ä–∞–∑–≤–∏—Ç–∏–µ–º –∏–Ω–Ω–æ–≤–∞—Ü–∏–æ–Ω–Ω–æ–≥–æ –ø—Ä–æ–¥—É–∫—Ç–∞',
					lection: [2, 3, 4, 5, 6, 7],
					practical: [8, 9],
					description: '–î–µ–≥—Ç—è—Ä—ë–≤–∞ –í.–í.',
					audience: '–õ–ö - 402',
				},
				{
					start: '2024-01-22T11:35:00',
					end: '2024-01-22T13:05:00',
					title: '–£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ —Ä–∞–∑–≤–∏—Ç–∏–µ–º –∏–Ω–Ω–æ–≤–∞—Ü–∏–æ–Ω–Ω–æ–≥–æ –ø—Ä–æ–¥—É–∫—Ç–∞',
					lection: [],
					practical: [2, 3, 4, 5, 6, 7, 8, 9],
					description: '–î–µ–≥—Ç—è—Ä—ë–≤–∞ –í.–í.',
					audience: '–õ–ö - 402',
				},
				{
					start: '2024-01-22T13:45:00',
					end: '2024-01-22T15:15:00',
					title: '–£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ —Ä–∞–∑–≤–∏—Ç–∏–µ–º –∏–Ω–Ω–æ–≤–∞—Ü–∏–æ–Ω–Ω–æ–≥–æ –ø—Ä–æ–¥—É–∫—Ç–∞',
					lection: [],
					practical: [3, 5, 7],
					description: '–î–µ–≥—Ç—è—Ä—ë–≤–∞ –í.–í.',
					audience: '–õ–ö - 402',
				},
			],
		},
	]
	it('Works for weekday', () => {
		expect(getLessonsForDay(WEEK_DAY, MOCK_SCHEDULE_DATA, { weekOffset: false }, 4)).toMatchSnapshot()
	})
})
