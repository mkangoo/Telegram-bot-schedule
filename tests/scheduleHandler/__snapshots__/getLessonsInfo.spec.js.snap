// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Works when today is weekday > Works for weekday 1`] = `
"

⏰ 09:55 - 11:25
Лекция:
<b>Управление развитием инновационного продукта</b>
Дегтярёва В.В.
ЛК - 402 

⏰ 11:35 - 13:05
Практика:
<b>Управление развитием инновационного продукта</b>
Дегтярёва В.В.
ЛК - 402"
`;
